### ISSUE TITLE
Testing in a production environment give us problems activating onlyoffice.documentserver.accessOnly=true

### ISSUE THREAD

**Issue by extrememicro** [exo-addons/onlyoffice#3]
Onlyoffice fails to get file:
```
2017-10-16 10:03:30,928 | DEBUG | > Onlyoffice document status: me@0e6b27cb-6bfe-3d9c-ab39-10910a09fe9d {"key":"0e6b27cb-6bfe-3d9c-ab39-10910a09fe9d","status":1,"users":["me"],"actions":[{"type":1,"userid":"me"}]} from 127.0.0.1(EXO_EXTERNAL_PUBLIC_IP) [o.e.onlyoffice.rest.EditorService<http-nio-127.0.0.1-8080-exec-8>] 
2017-10-16 10:03:30,928 | WARN  | Attempt to update status by not allowed host: 127.0.0.1 [o.e.onlyoffice.rest.EditorService<http-nio-127.0.0.1-8080-exec-8>] 
2017-10-16 10:03:30,954 | INFO  | Adding the key: F26570740F25DC56B3C17B455C18DDE4 [SessionCreatedListener<http-nio-127.0.0.1-8080-exec-25>] 
2017-10-16 10:03:30,956 | DEBUG | > Onlyoffice document content: me@0e6b27cb-6bfe-3d9c-ab39-10910a09fe9d to 127.0.0.1(EXO_EXTERNAL_PUBLIC_IP) [o.e.onlyoffice.rest.EditorService<http-nio-127.0.0.1-8080-exec-25>] 
2017-10-16 10:03:30,956 | WARN  | Attempt to download content by not allowed host: 127.0.0.1 [o.e.onlyoffice.rest.EditorService<http-nio-127.0.0.1-8080-exec-25>] 
2017-10-16 10:03:31,964 | INFO  | Adding the key: A828B3ABA14AB4C007E312121197DDA5 [SessionCreatedListener<http-nio-127.0.0.1-8080-exec-3>] 
2017-10-16 10:03:31,966 | DEBUG | > Onlyoffice document content: me@0e6b27cb-6bfe-3d9c-ab39-10910a09fe9d to 127.0.0.1(EXO_EXTERNAL_PUBLIC_IP) [o.e.onlyoffice.rest.EditorService<http-nio-127.0.0.1-8080-exec-3>] 
2017-10-16 10:03:31,966 | WARN  | Attempt to download content by not allowed host: 127.0.0.1 [o.e.onlyoffice.rest.EditorService<http-nio-127.0.0.1-8080-exec-3>] 
```
We have exo and onlyoffice server in separate public ips, reachable to internet and using https. we had nginx reverse proxy running in both servers.

It's anormal that clientIp was exo external public ip/127.0.0.1.
In exo.properties onlyoffice.documentserver.host was a different one.

Regards

**Comment by extrememicro** [comment]
I forgot to mention that we had nginx reverse proxy in both servers.

**Comment by pnedonosko** [comment]
Point in _exo.properties_ in _onlyoffice.documentserver.host_ a host name which actually will be used, use of localhost isn't supported - this check is simply a name string comparison.

**Comment by pnedonosko** [comment]
If nginx proxy runs on the same host as the Platform server, you still have an IP of this host. A parameter _onlyoffice.documentserver.host_ on eXo side, it's an IP of the Document Server that will request eXo REST service when document will be updated by Onlyoffice editor. This DS IP, if not running at the same host as eXo, will be definitely not localhost - use an IP that appears after the proxy, which will be an effective request IP on eXo side.

**Comment by extrememicro** [comment]
onlyoffice.documentserver.host=somedomain.someweb.com
onlyoffice.documentserver.schema=https
onlyoffice.server.host=someotherodmain.someotherweb.com
onlyoffice.server.schema=https
onlyoffice.documentserver.accessOnly=true

All two servers are different. I think that the problem is with getClientIpAddr getting the wrong ip becasue nginx reverse proxy.

**Comment by pnedonosko** [comment]
It looks that you need setup your nginx to allow seeing real remote IP addresses, like discussed [here](https://serverfault.com/questions/656616/set-remote-addr-to-real-client-ip) or may be [here](https://serverfault.com/questions/314574/nginx-real-ip-header-and-x-forwarded-for-seems-wrong).

**Comment by extrememicro** [comment]
No. The problem was with an incorrect x-headers use in getClientHost().
X-Forwarded-Host is the original host requested by the client in the Host HTTP request header. Usually null or exo server host name/ip. Therefore, this header value is not desired to do the string comparison here: documentserverHostName.equals(hostName) 
Finally, we saw the use of getRemoteHost() to try to get onlyoffice host name for the later comparison but if host name is empty, then it tries to retrieve the IP address and it will be the last proxy if there is one involved. A IP address will always fail in the documentserverHostName comparison.

... so a solution could be to use X-Forwarded-For (IP address of the client) and get host name from it. For this to work, in server.xml, enableLookups="true" is needed for request.getRemoteHost() to perform DNS lookups in order to return the actual host name of the remote client. 

We test this here and all is fine:
https://github.com/extrememicro/onlyoffice/commit/3b1892fa35c6946f434e02d7f39b9adb9d90a447

Said that, we think is much better to use tokens like onlyoffice docs expose here: https://api.onlyoffice.com/editors/security

**Comment by pnedonosko** [comment]
Ok, your findings change the picture. First it would be a quick solution to rely on "X-Forwarded-For" header as you did test and it was fine (we'll add it to 1.1.x codebase). 
Security tokens in Onlyoffice it's something new they've added recently, sure it sounds as a feature to use and it should be tried for a next minor release of the addon (1.2.x codebase).

**Comment by pnedonosko** [comment]
Try set _onlyoffice.documentserver.host_ with an IP, not host name, of your Document Server. Does it fix the problem for you?

**Comment by extrememicro** [comment]
Yes we tried it. Check will pass and exo and DS will communicate but user will not connect to DS.
The problem with putting an ip in a production environment is that it will not resolve to document server because nginx can't serve an unknown domain.

**Comment by pnedonosko** [comment]
@extrememicro have you solved the problem on your side? Can you share a final solution (proxy setup or another workaround you did)?

**Comment by pnedonosko** [comment]
@extrememicro is this problem still actual with latests commits?
